### PR TITLE
Implement handling of encoded block hash for storage RPC functions

### PR DESCRIPTION
**Pull request by arijitAD**
## Changes

- Handle block hash in `GetRuntimeVersion()` `GetMetadata()` `GetPairs()`

## Tests

```
TestStateRPCAPI()
```

## Checklist

- [x] I have read [CODE_OF_CONDUCT](https://github.com/ChainSafe/gossamer/blob/development/.github/CODE_OF_CONDUCT.md) and [CONTRIBUTING](https://github.com/ChainSafe/gossamer/blob/development/.github/CONTRIBUTING.md) 
- [x] I have provided as much information as possible and necessary
- [x] I have reviewed my own pull request before requesting a review
- [x] All integration tests and required coverage checks are passing

## Issues

- #834

### CODE REVIEW

File: dot/rpc/modules/state.go [hunk 1]
```diff
@@ -126,11 +126,19 @@ func NewStateModule(net NetworkAPI, storage StorageAPI, core CoreAPI) *StateModu
...
+
+	// Extract the params
 	pReq := *req
 	reqBytes, _ := common.HexToBytes(pReq[0])
+	bhash, _ := common.HexToHash(pReq[1])
+
+	stateRootHash, err := sm.storageAPI.GetStateRootFromBlock(bhash)
+	if err != nil {
+		return err
+	}
```
Comment (noot) [inline dot/rpc/modules/state.go:139]: I think the block hash parameter is optional (ie. if `len(pReq) == 1` then there is no block hash provided), if no block hash is provided, then the latest data is used which is what was happening before. could you add a case to check `len(pReq)` to see if you should provide a block hash or not?

File: dot/rpc/modules/state.go [hunk 2]
```diff
@@ -188,10 +196,12 @@ func (sm *StateModule) GetKeys(r *http.Request, req *StateStorageKeyRequest, res
 // GetMetadata calls runtime Metadata_metadata function
 func (sm *StateModule) GetMetadata(r *http.Request, req *StateRuntimeMetadataQuery, res *string) error {
 	// TODO implement change storage trie so that block hash parameter works (See issue #834)
-	metadata, err := sm.coreAPI.GetMetadata()
+	bhash := common.Hash(*req)
+	metadata, err := sm.coreAPI.GetMetadata(bhash)
```
Comment (noot) [inline dot/rpc/modules/state.go:200]: similar here, I believe it's optional

File: dot/rpc/modules/state.go [hunk 3]
```diff
@@ -201,7 +211,8 @@ func (sm *StateModule) GetMetadata(r *http.Request, req *StateRuntimeMetadataQue
 //  If no block hash is provided, the latest version gets returned.
 // TODO currently only returns latest version, add functionality to lookup runtime by block hash (see issue #834)
 func (sm *StateModule) GetRuntimeVersion(r *http.Request, req *string, res *StateRuntimeVersionResponse) error {
-	rtVersion, err := sm.coreAPI.GetRuntimeVersion()
+	bhash, _ := common.HexToHash(*req)
+	rtVersion, err := sm.coreAPI.GetRuntimeVersion(bhash)
```
Comment (noot) [inline dot/rpc/modules/state.go:215]: same here

File: dot/state/storage.go [hunk 4]
```diff
@@ -203,6 +203,15 @@ func (s *StorageState) GetStorageByBlockHash(bhash common.Hash, key []byte) ([]b
 	return s.GetStorage(&header.StateRoot, key)
 }
 
+// GetStateRootFromBlock returns the state root hash of a given block hash
+func (s *StorageState) GetStateRootFromBlock(bhash common.Hash) (*common.Hash, error) {
+	header, err := s.blockState.GetHeader(bhash)
+	if err != nil {
+		return nil, err
+	}
+	return &header.StateRoot, nil
```
Comment (noot) [inline dot/state/storage.go:212]: is there a reason it returns a pointer?

Comment (Tikaryan) [reply dot/state/storage.go:212]: blockState.GetHeader() above returns a pointer and we later on user this as pointer in other functions.

Comment (arijitAD) [reply dot/state/storage.go:212]: `blockState.GetHeader()` above returns a pointer, and we use it as a pointer in other functions later on.

**Review by noot — APPROVED** [review]
looks good to me! if possible, could you add unit tests for the state functions, for code coverage?

**Comment by arijitAD** [comment]
> looks good to me! if possible, could you add unit tests for the state functions, for code coverage?

Added unit test.

**Comment by RyRy79261** [comment]
@arijitAD 

The test is crashing on my machine: 
```Running tool: /usr/bin/go test -timeout 30s -run ^(TestStateRPC|TestStateRPCAPI)$ github.com/ChainSafe/gossamer/tests/rpc

Going to start RPC suite test
Going to skip RPC suite tests
t=2020-11-30T20:18:38+0200 lvl=info msg="initializing gossamer..." pkg=test/utils cmd="/home/ryann/Documents/Repos/ChainSafe/gossamer/bin/gossamer init --config /home/ryann/Documents/Repos/ChainSafe/gossamer/tests/utils/config_babe_max_threshold.toml --basepath /tmp/TestStateRPCAPI/alice --genesis-raw /home/ryann/Documents/Repos/ChainSafe/gossamer/chain/gssmr/genesis-raw.json --force"
t=2020-11-30T20:18:39+0200 lvl=info msg="initialized gossamer!" pkg=test/utils node=0
t=2020-11-30T20:18:39+0200 lvl=info msg="starting gossamer..." pkg=test/utils cmd="/home/ryann/Documents/Repos/ChainSafe/gossamer/bin/gossamer --port 7000 --config /home/ryann/Documents/Repos/ChainSafe/gossamer/tests/utils/config_babe_max_threshold.toml --basepath /tmp/TestStateRPCAPI/alice --rpchost  --rpcport 8540 --rpcmods system,author,chain,state,dev --rpc --log info --roles 4 --key alice"
t=2020-11-30T20:18:40+0200 lvl=info msg="node started" pkg=test/utils key=alice cmd.Process.Pid=122339
--- FAIL: TestStateRPCAPI (8.86s)
    /home/ryann/Documents/Repos/ChainSafe/gossamer/tests/rpc/rpc_05-state_test.go:117: starting gossamer...
    --- FAIL: TestStateRPCAPI/Test_invalid_block_hash_state_getRuntimeVersion (0.02s)
        rpc_05-state_test.go:223: 
            	Error Trace:	rpc_05-state_test.go:223
            	Error:      	"{\"jsonrpc\":\"2.0\",\"result\":{\"specName\":\"node\",\"implName\":\"substrate-node\",\"authoringVersion\":10,\"specVersion\":193,\"implVersion\":193,\"apis\":[[\"0xdf6acb689907609b\",2],[\"0x37e397fc7c91f5e4\",1],[\"0x40fe3ad401f8959a\",3],[\"0xd2bc9897eed08f15\",1],[\"0xf78b278be53f454c\",1],[\"0xed99c5acb25eedf5\",2],[\"0xcbca25e39f142387\",1],[\"0xbc9d89904f5b923f\",1],[\"0x68b66ba122c93fa7\",1],[\"0x37c8bb1350a9a2a8\",1],[\"0xab3c0572291feb8b\",1]]},\"id\":1}\n" does not contain "Key not found"
            	Test:       	TestStateRPCAPI/Test_invalid_block_hash_state_getRuntimeVersion
    --- FAIL: TestStateRPCAPI/Test_invalid_block_hash_state_getPairs (0.07s)
        rpc_05-state_test.go:223: 
            	Error Trace:	rpc_05-state_test.go:223
            	Error:      	
            	Test:       	TestStateRPCAPI/Test_invalid_block_hash_state_getPairs
    --- FAIL: TestStateRPCAPI/Test_invalid_block_hash_state_getMetadata (0.00s)
        rpc_05-state_test.go:223: 
            	Error Trace:	rpc_05-state_test.go:223
            	Error:      	"{\"jsonrpc\":\"2.0\",\"error\":{\"code\":-32600,\"message\":\"json: cannot unmarshal string into Go value of type modules.StateRuntimeMetadataQuery\",\"data\":[\"0x580d77a9136035a0bc3c3cd86286172f7f81291164c5914266073a30466fba21\"]},\"id\":1}\n" does not contain "Key not found"
            	Test:       	TestStateRPCAPI/Test_invalid_block_hash_state_getMetadata
    --- FAIL: TestStateRPCAPI/Test_required_params_missing_hash_state_getPairs (0.00s)
        rpc_05-state_test.go:220: 
            	Error Trace:	rpc_05-state_test.go:220
            	Error:      	Expected nil, but got: &url.Error{Op:"Post", URL:"http://:8540", Err:(*errors.errorString)(0xc00007e040)}
            	Test:       	TestStateRPCAPI/Test_required_params_missing_hash_state_getPairs
    /home/ryann/Documents/Repos/ChainSafe/gossamer/tests/rpc/rpc_05-state_test.go:126: going to tear down gossamer...
FAIL
FAIL	github.com/ChainSafe/gossamer/tests/rpc	8.877s
FAIL
```

**Comment by arijitAD** [comment]
> @arijitAD
> 
> The test is crashing on my machine:
> 
> ```
> 
> Going to start RPC suite test
> Going to skip RPC suite tests
> t=2020-11-30T20:18:38+0200 lvl=info msg="initializing gossamer..." pkg=test/utils cmd="/Documents/Repos/ChainSafe/gossamer/bin/gossamer init --config usr/Documents/Repos/ChainSafe/gossamer/tests/utils/config_babe_max_threshold.toml --basepath /tmp/TestStateRPCAPI/alice --genesis-raw usr/Documents/Repos/ChainSafe/gossamer/chain/gssmr/genesis-raw.json --force"
> t=2020-11-30T20:18:39+0200 lvl=info msg="initialized gossamer!" pkg=test/utils node=0
> t=2020-11-30T20:18:39+0200 lvl=info msg="starting gossamer..." pkg=test/utils cmd="usr/Documents/Repos/ChainSafe/gossamer/bin/gossamer --port 7000 --config usr/Documents/Repos/ChainSafe/gossamer/tests/utils/config_babe_max_threshold.toml --basepath /tmp/TestStateRPCAPI/alice --rpchost  --rpcport 8540 --rpcmods system,author,chain,state,dev --rpc --log info --roles 4 --key alice"
> t=2020-11-30T20:18:40+0200 lvl=info msg="node started" pkg=test/utils key=alice cmd.Process.Pid=122339
> --- FAIL: TestStateRPCAPI (8.86s)
>     usr/Documents/Repos/ChainSafe/gossamer/tests/rpc/rpc_05-state_test.go:117: starting gossamer...
>     --- FAIL: TestStateRPCAPI/Test_invalid_block_hash_state_getRuntimeVersion (0.02s)
>         rpc_05-state_test.go:223: 
>             	Error Trace:	rpc_05-state_test.go:223
>             	Error:      	"{\"jsonrpc\":\"2.0\",\"result\":{\"specName\":\"node\",\"implName\":\"substrate-node\",\"authoringVersion\":10,\"specVersion\":193,\"implVersion\":193,\"apis\":[[\"0xdf6acb689907609b\",2],[\"0x37e397fc7c91f5e4\",1],[\"0x40fe3ad401f8959a\",3],[\"0xd2bc9897eed08f15\",1],[\"0xf78b278be53f454c\",1],[\"0xed99c5acb25eedf5\",2],[\"0xcbca25e39f142387\",1],[\"0xbc9d89904f5b923f\",1],[\"0x68b66ba122c93fa7\",1],[\"0x37c8bb1350a9a2a8\",1],[\"0xab3c0572291feb8b\",1]]},\"id\":1}\n" does not contain "Key not found"
>             	Test:       	TestStateRPCAPI/Test_invalid_block_hash_state_getRuntimeVersion
>     --- FAIL: TestStateRPCAPI/Test_invalid_block_hash_state_getPairs (0.07s)
>         rpc_05-state_test.go:223: 
>             	Error Trace:	rpc_05-state_test.go:223
>             	Error:      	
>             	Test:       	TestStateRPCAPI/Test_invalid_block_hash_state_getPairs
>     --- FAIL: TestStateRPCAPI/Test_invalid_block_hash_state_getMetadata (0.00s)
>         rpc_05-state_test.go:223: 
>             	Error Trace:	rpc_05-state_test.go:223
>             	Error:      	"{\"jsonrpc\":\"2.0\",\"error\":{\"code\":-32600,\"message\":\"json: cannot unmarshal string into Go value of type modules.StateRuntimeMetadataQuery\",\"data\":[\"0x580d77a9136035a0bc3c3cd86286172f7f81291164c5914266073a30466fba21\"]},\"id\":1}\n" does not contain "Key not found"
>             	Test:       	TestStateRPCAPI/Test_invalid_block_hash_state_getMetadata
>     --- FAIL: TestStateRPCAPI/Test_required_params_missing_hash_state_getPairs (0.00s)
>         rpc_05-state_test.go:220: 
>             	Error Trace:	rpc_05-state_test.go:220
>             	Error:      	Expected nil, but got: &url.Error{Op:"Post", URL:"http://:8540", Err:(*errors.errorString)(0xc00007e040)}
>             	Test:       	TestStateRPCAPI/Test_required_params_missing_hash_state_getPairs
>     usr/Documents/Repos/ChainSafe/gossamer/tests/rpc/rpc_05-state_test.go:126: going to tear down gossamer...
> FAIL
> FAIL	github.com/ChainSafe/gossamer/tests/rpc	8.877s
> FAIL
> ```

It was not getting the `HOSTNAME` from env. Fixed it.

**Review by RyRy79261 — COMMENTED** [review]
I think theres some unneeded, commited test data in here:
`dot/rpc/modules/test_data/expected_metadata`

**Comment by arijitAD** [comment]
> I think theres some unneeded, commited test data in here:
> `dot/rpc/modules/test_data/expected_metadata`

The data was included in the test which made it difficult to read. I stored the binary data in a separate file to make the test more readable.

**Review by RyRy79261 — APPROVED** [review]

**Pull request merged**